lib/archethic/transaction_chain/transaction/validation_stamp/ledger_operations.ex

```diff
@@ -424,8 +420,8 @@ defmodule Archethic.TransactionChain.Transaction.ValidationStamp.LedgerOperation
       {{:token, token_address, token_id}, inputs} ->
         token_used?(tokens_to_spend, token_address, token_id) or consolidate_inputs?(inputs)
 
-      {:state, _} ->
-        encoded_state != nil
+      {:state, [%UnspentOutput{encoded_payload: previous_state}]} ->
+        encoded_state != previous_state
 
       {:call, inputs} ->
         case contract_context do
```

Comment:
It looks like there is an invalid filter function here, filter should receive true or false, but here it can receive nil, an utxo or an empty list.
Also, if the return is an empty list or an utxo, the filter will be true, and then we will consume all the inputs of type call, but we want to consume only the call that has been used.
So maybe the filter could be replaced by a flat_map directly and all the function would return a list with the utxo to consume.

Reply:
The state in this case can only exist with a single unspent output. It cannot be nil or empty list as there is only one state utxo. 
Hence the nil cannot be present as it's not the encoded state we use in the group by

Reply:
I was talking about call utxo, not state

Reply:
Ah... yes, indeed it's not well used. 
It can be improved accordingly.